### PR TITLE
OIDC authentication against API server: allow multiple values in token for requiredClaims

### LINKED ISSUE #101291: Required claims does not support arrays and failing kubernetes auth with oidc - EKS
We have added an IDP called "lmiprod" to the EKS cluster with the following command and it has requiredClaims which specifies roles as a list and everyone in the list who has access prod role should be able to access the cluster and currently they are receiving the unauthorized error with requiredClaims but without that everyone in the OIDC app is able to access the cluster.

----
aws eks associate-identity-provider-config --cluster-name xxxx-1 --oidc 'identityProviderConfigName=lmiprod,issuerUrl=https://login.microsoftonline.com/<xxxx>/v2.0,clientId=ClientId,usernameClaim=email,usernamePrefix=oidc:,groupsClaim=groups,groupsPrefix=oidc:,requiredClaims={roles=prod}' --region us-west-2

The token which we are trying to use looks like this:

{
  "aud": "xxx",
  "iss": "https://login.microsoftonline.com/ffe3c2ff-f367-4962-b7ee-6e7bba609fb8/v2.0",
  "iat": 1618472379,
  "nbf": 1618472379,
  "exp": 1618476279,
  "email": "xxxxx",
  "groups": [
    "xxxx",
    "xxxx",
    "xxxx"
  ],
  "name": "xxxx",
  "nonce": "xxxx",
  "oid": "xxxx",
  "preferred_username": "xxxx",
  "rh": "xxxxx",
  "roles": [
    "rxxxx",
    "dxxx",
    "pxxxx"
  ],
  "sub": "xxxxxx",
  "tid": "xxxxxx",
  "uti": "xxxxxx",
  "ver": "2.0"
}

So the "roles" claim is a list and we want to allow everyone that has the "prod" role to access the cluster. This doesn't work, 

we get back:
-------
I0415 10:16:40.885213 2190013 round_trippers.go:445] GET https://xxxxx.sk1.us-west-2.eks.amazonaws.com/api?timeout=32s 401 Unauthorized in 319 milliseconds
I0415 10:16:40.885240 2190013 round_trippers.go:451] Response Headers:
I0415 10:16:40.885260 2190013 round_trippers.go:454]     Date: Thu, 15 Apr 2021 08:16:40 GMT
I0415 10:16:40.885269 2190013 round_trippers.go:454]     Audit-Id: 19ef2f74-52c0-43b0-9d0e-b8338256efa7
I0415 10:16:40.885276 2190013 round_trippers.go:454]     Cache-Control: no-cache, private
I0415 10:16:40.885284 2190013 round_trippers.go:454]     Content-Type: application/json
I0415 10:16:40.885291 2190013 round_trippers.go:454]     Content-Length: 129
I0415 10:16:40.888998 2190013 request.go:1107] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Unauthorized","reason":"Unauthorized","code":401}
I0415 10:16:40.892985 2190013 cached_discovery.go:121] skipped caching discovery info due to Unauthorized
I0415 10:16:40.893256 2190013 helpers.go:216] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}]

Not specifying the "requiredClaims" field works, but allows all people in that OIDC app to access the cluster.

Arrays are supported with groups but not with requiredClaims:

For more information:
----
Array support is present only for group claims - https://github.com/kubernetes/kubernetes/blob/27cf50d85edb91357d56fd762271974e7a7254bc/staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go#L600-L610

Required claims does not support arrays - https://github.com/kubernetes/kubernetes/blob/27cf50d85edb91357d56fd762271974e7a7254bc/staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go#L626

### PR DESCRIPTION
**Pull request by alfredkrohmer**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

When using `requiredClaims`, currently only sting values in the OIDC token are accepted. We want to use this so that we can have an array of values in the token and the value in `requiredClaims` must be in that array.

Example:
OIDC token contains claim `roles` with values `["dev", "prod"]`. We want to be able to specify `requiredClaims` as `roles: dev` so that any user that has the `dev` role can be authenticated.

#### Which issue(s) this PR fixes:

Fixes  #101291

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by enj — CHANGES_REQUESTED** [review]
Overall I am unsure about this change as it changes the meaning of the existing check.

`foo == val` and `foo in [val1 val2 val3]` are not exactly the same thing.

The existing flag based config is really terrible, and I do not want to expand it further but this seems dangerous without opting in somehow.  This also opens the door to wanting stuff like `both foo and bar in [val1, val2, ...]`.  I would like to see this stuff more holistically handled in a proper config format.

/hold

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 1]
```diff
@@ -611,23 +611,27 @@ func (a *Authenticator) AuthenticateToken(ctx context.Context, token string) (*a
...
-		if err := c.unmarshalClaim(claim, &claimValue); err != nil {
+		var claimValues stringOrArray
+		if err := c.unmarshalClaim(claim, &claimValues); err != nil {
 			return nil, false, fmt.Errorf("oidc: parse claim %s: %v", claim, err)
 		}
-		if claimValue != value {
-			return nil, false, fmt.Errorf("oidc: required claim %s value does not match. Got = %s, want = %s", claim, claimValue, value)
+		for _, claimValue := range []string(claimValues) {
+			if claimValue == value {
+				return resp, true, nil
```
[the remark this answers is not in the document]
Comment (devkid) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Fixed

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 2]
```diff
@@ -611,23 +611,27 @@ func (a *Authenticator) AuthenticateToken(ctx context.Context, token string) (*a
 		}
 	}
 
+	resp := &authenticator.Response{User: info}
+
```
[the remark this answers is not in the document]
Comment (devkid) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: Fixed

**Pull request closed without merging**